### DISCUSSION TITLE
Should we split integration libraries into separate repositories?

### DISCUSSION THREAD

**Discussion by aivascu** [AutoFixture/AutoFixture#1260] [Ideas]
Currently whenever we are making changes to one of the libraries we are forced to release new versions for all other packages, which might not have changed since the last release nor do they need newer versions of the main package.
To solve this issue we could split up the source for integration libraries (AutoMoq, AutoNSubstitute, etc.) into separate repositories.

## Benefits
- Integration libraries will be able to iterate more often and independently from the main `AutoFixture` package
  - Faster growing integration libraries (e.g. AutoFakeItEasy) will be able to introduce breaking changes and deprecate older API
  - Less conditional dependencies
- Integration libraries will depend on the earliest AutoFixture package version necessary
- End users will receive meaningful updates
- More focused issues and PRs raised (current repo can still act as a hub for the community)
- Smaller solution size (Visual Studio will have an easier time)
- Faster builds

## Drawbacks
- Harder to debug the integration libraries end-to-end
- More infrastructure code: build scripts, readme updates, etc. (doesn't change as often though)

**Comment by Kralizek** [comment]
I second this idea :) It's a lot of work but I think it's worth.

**Comment by aivascu** [comment]
@zvirja @ecampidoglio  @moodmosaic @adamchester do you have any thoughts on this?

**Reply by moodmosaic** [reply]
You'll be maintaining separare projects once you split them... You already made a good analysis above. 

My main concern is that maintaining separare projects probably requires more effort compared to a monorepo.

**Comment by zvirja** [comment]
I would personally advocate to continue with a mono-repo we have today. Here are my arguments:

## AutoFixture & glue libraries is a single product

Even though AutoFixture Core could be used independently of glue libraries, I've never seen that in my experience. Usually people install it together with a glue library for the test framework. Also when we refer to AutoFixture product, we usually imagine it all as a single thing. SemanticComparison on the other hand, is a completely independent and standalone project.

## Easier for clients to deal with packages

Typically when creating a test project we have to import test framework glue lib + mock glue lib + often `Idioms`. As of today it's very easy to align version usage by just having the same package number everywhere. That's what our clients (including myself) learned and that's what makes it simple.

After introducing independent versioning to each lib customers will have to spend more efforts on synchronizing lib versions and tackling with ranges of compatible versions (e.g. bumping up NSubstitute version and realizing that now xUnit is incompatible and should be updated as well).

## Glue libraries are always compatible with latest AutoFixture

As of today each latest glue library is compatible with latest `AutoFixture`. After we split each lib to a standalone repo, I imagine we depend on `AutoFixture` by specifying a range with an upper boundary (e.g. `[4.0-5.0)`). If we do not do so, we end up in a glue library not being actually compatible with future versions of core and misbehaving. If we do it (what we should), after each major release of AutoFixture Core we'll have to create 15 PRs to make sure all glue libraries can consume it.

As of today we solve that by simply building all the projects together.

## Mono repo keep libs in more consistent state

Currently when we introduce a feature e.g. to `AutoFixture.NUnit`, we are considering at the same time whether we have to port the same change to other glue libraries. We could do all the changes via a single PR, making all the glue libs consistent.

After we shift each glue library to a standalone repo,  it would be easy to skip a fact that we have to align change with other glue libs. The repo structure will imply that each glue library is an independent project, rather than being a part of the same product. Even if we remember, we will have to ask people to create 3 more PRs to other projects which they will never do.

## Mono repo makes cross-project changes easier

If we need to introduce a breaking change (in major version) or just introduce a feature requiring addition in core, we can do all of it in a single PR, rather than having a need to scatter & sync PRs among multiple repos.

# Conclusions

Given the aforementioned I would advise to keep everything within a single repo. I do realize that each approach is a trade-off, but to me the price we pay today is way lesser than when product is scattered. And especially giving that development of this product is almost dead and we just maintain the project, the model with mono repo works well.

Redundant releases, longer build time, slow IDE startup (which is mitigated by opening specific solutions rather than `All`) - that's a very small inconvenience to me compared to hassle of synchronizing among repos and having to keep a consistent look and feel.

That said though I would like to also emphasize that it's my opinion and analysis only. If you have time, energy and vision to split the project, see clear solution to the points I raised above - I do not want you to feel blocked by inactive maintainers of this project who do not contribute actively anyway (like I felt when I initially joined the project). Just make sure it really worth it both for you and for our clients 😉

**Comment by ecampidoglio** [comment]
The key thing about the glue libraries is that you have to make sure that they keep working with every new change made to AutoFixture. By far the easiest way to do that is by having them in the same solution.

If we were to split them into separate repos, we'd have to include the AutoFixture repo as a submodule in order to be able to do that. While that's certainly possible, it adds significant management overhead.

Is it worth it?

Well, I personally don't think the build times are an issue. I do see an advantage in being able to update the libraries independently from each other, because they often have different reasons for requiring an update (e.g. a new version of `Moq` shouldn't require an update to `AutoFixture` or `AutoFixture.Xunit2`). I also see the advantage in compartmentalizing issues and discussions around different libraries by having them in separate repos.

With that said, maintaining compatibility between the libraries and AutoFixture is paramount and I don't think that the cost of dealing with submodules (nor NuGet dependency hell) is worth the advantages I listed above, especially since we control all the libraries ourselves.

**tl;dr** I vote for staying with the single repo approach.

**Reply by moodmosaic** [reply]
Agreed.

**Comment by aivascu** [comment]
@moodmosaic @zvirja @ecampidoglio thank you for the feedback. I appreciate you taking the time to share your views on the topic. 
Taking into consideration all that has been said, I believe the agreement is to continue developing AutoFixture as a monorepo.
